### ISSUE TITLE
Reduce jquery's size

### ISSUE THREAD

**Issue by franciscop** [jquery/jquery#2892]
TL;DR, I would love to help reducing jquery's size while keeping the same performance and compatibility

I am the creator of [Umbrella JS](http://umbrellajs.com/), made as a lightweight jquery alternative for modern browsers. I started javascript with jquery many years ago and I'd love to contribute back few things I've learned to try to reduce jquery's size.

While reading jquery's code I've realized that there's many things made in that way in the name of performance; however modern browsers and engines are very different. Take for example the `while()` loop vs `.forEach()` or `.reduce()`, that is traditionally considered a performance improvement:

```js
var cur = "", i = 0, c;
while ( ( c = arr[ i++ ] ) ) {
  if ( cur.indexOf( " " + c + " " ) < 0 ) {
    cur += c + " ";
  }
}
```

vs

```js
// This can be generalized and moved out
function joinClassNames(prev, name){
  if (prev.indexOf(" " + name + " ") < 0) {
    prev += name + " ";
  }
  return prev;
};
var cur = arr.reduce(joinClassNames, "");
```

However, now both of these methods are at the same speed (within margin error) as jsperf shows: **[jsperf test](http://jsperf.com/while-vs-foreach-array)**.

While they are similar speed, the second one could really help [addClass](https://github.com/jquery/jquery/blob/a4d16a26abe3ea36337f7bfc4d1bfe13b893cd17/src/attributes/classes.js) and other `*Class` methods, since the function itself could be put at the beginning (similarly to `getClass()`) and get reused, slimming down the code.

With all this said, I'd love to try to contribute back cleaning, reducing and reusing as much source code as possible in jquery. I have a couple of questions though:

- **Is this a wanted improvement?** The last thing I'd want is to waste jquery's developers time and my time doing something that no one calling the shots is concerned about.
- **How do I test jquery's performance?** Are there currently performance tests? I'm willing to try to add some, but I'd need some help setting this up. Manually testing each and every change in [jsperf](http://jsperf.com/) is not viable (not reproducible). I [have some setup in Umbrella JS](https://github.com/umbrellajs/umbrella/blob/master/src/test.js#L64) to run in Circle CI.

PS, there was some discussion going on [in issue #2887](https://github.com/jquery/jquery/issues/2887)

**Comment by markelog** [comment]
We would **love** for you to contribute code that would increase perf and reduce size! Also, if you want, you can reach us in irc freenode with jquery-dev channel.

> How do I test jquery's performance?

Mainly we use jsperf and sometimes manual profiling. But we were talking for years now to create a reliable benchmark suite with graphs and all, so every commit would show us how performance has changed on every commit (like [plato](https://github.com/es-analysis/plato) but for perf). 

Second way, is to integrate benchs to test-suite with the cap, like on your project. 

Third is to create lodash-like [bench](https://github.com/lodash/lodash/blob/b2984bd034b8774ee5a4ed31c6798e66aa6f4dc9/perf/perf.js) comparing previous jQuery versions or like https://github.com/shama/grunt-benchmark

Also in sizzle we have [this](https://github.com/jquery/sizzle/tree/a7bbfc9a7195b9bd32f03a365d25dc9a55d6f5bd/speed), it also requires manual execution, but it would be much cooler to execute it with saucelabs or browserstack. In core we had something similar but those tests were much outdated.

**Comment by gibson042** [comment]
Also note that when comparing size, it is the gzipped minified file that we care about (and what's smaller before compression isn't necessarily smaller after compression), and that we generally favor size over performance (except in particularly hot paths or when the difference is exceptionally stark).

**Comment by franciscop** [comment]
@markelog I am trying to reduce size, not increase performance (but I was asking about performance tests not to decrease it). However I realized that the test suite is completely different to what I knew so I need first to understand how it works before I can even dream of adding anything like graphics over time. A reliable benchmark would be really good, but sometimes it's difficult because it's opinionated. Do you optimize for `.addClass('test')` or for `.addClass(function(){ return 'test'; });`? (just an example)

@gibson042 I didn't know that size was more important than performance, so I'll take it into account. I am checking the gzipped that I see from the `grunt compare_size`. I know that literal repetitions take almost no space in gzip while unique character strings (like error messages) take a lot, but most of the times what will result in a smaller or bigger gzipped file is a mistery. If you have any more tips I'd be really thankful.

**Comment by dmethvin** [comment]
It's always good to have some new people reviewing and rethinking the code! We don't have any comprehensive performance tests, we would love to have some but it's just a lot of work to create them. As far as what we optimize for, we have to guess about the usage of the APIs since we don't have any reliable way to gather actual usage. I would guess that adding classes by function is probably rare, and in any case it would be unusual that `.addClass()` was a performance bottleneck in real code.

One thing to keep in mind about benchmarks in jsperf is that if you are manipulating any DOM structures the time it takes to execute the code may be really small compared to the potential to force layout in the rendering engine. You see the results of that in areas like [showHide](https://github.com/jquery/jquery/blob/f1300f18877f0140994190bb1b3a945fa4b2fd3d/src/css/showHide.js#L32-L83) where we do all the DOM reading in one batch followed by the DOM changing.

Finally, keep in mind that we support a really broad set of browsers and not all of them support modern APIs. We don't want Android 2.3 or IE9 throwing an error and dying, although we may be willing to let them be "less right" in some cases as long as they don't blow up. You should be able to test browsers on BrowserStack or SauceLabs.

**Comment by markelog** [comment]
> However I realized that the test suite is completely different to what I knew so I need first to understand how it works before I can even dream of adding anything like graphics over time. 

Graphics and stuff would be an ideal result for me, but that is relatively hard and time consuming task, at this point, it is not necessary to shoot for stars, if we would have a tool integrated in CI build process which either would compare some numbers with the other ones or just checking the cap that would be already an awesome start.

In your project you integrate perf tests in test-suite, in our case, whereas we already automatically run our tests in supported browsers. So your paradigm might fit nicely.

> Do you optimize for .addClass('test') or for .addClass(function(){ return 'test'; });? (just an example)

Just like @dmethvin said, `.addClass('test')` will cause reflow in most browsers, that process is very expensive performance wise, comparing to the most hottest path we have in that method. Perf tests would be important for like `animation` method, like newest beta improved animation for some cases by 30fps - https://jsfiddle.net/vphby4ku/, which is huge win.

And echoing to the @gibson042 statement, i'd like to show you some example, we have [this](https://github.com/jquery/jquery/pull/2869) pull which should heavily improve the perf but increase the size.

That `jQuery.event.fix` is very expensive method which is always used in our event module, so it makes sense to accept that pull, since it would reflect on all events, if that change would indeed improve the perf that is. But on the other hand, if you will improve `addClass` by few rps (or 5% or something), by increasing the size - then that probably wouldn't be enough.

Counter example - i remember the time when i improved most traversing methods by 60%, but it wasn't landed because it added 50 bytes (or something like that), because traverse methods already fast enough. 

So following that experience, if you would provide the pull which would worsen the perf of traversing methods by 60% while decreasing the size then we will consider it, but in the same time, we probably wouldn't accept the pull which would decrease perf of event module by improving the size.

**Comment by timmywil** [comment]
Thank you for your willingness to help us improve jQuery! New contributors are always welcome, and we would appreciate fresh eyes on the codebase.

Please feel free to look over our code and submit as many pull requests as you deem appropriate. We will evaluate each of them with the criteria that we've always used; that is, based on size, performance, and readability.

I'll be closing this ticket not because I don't want to reduce size, but because we use our [Roadmap](https://github.com/jquery/jquery/wiki/Roadmap) to track issues that have no definite end (such as rewriting our internal speed framework). However, I won't be adding this one to the Roadmap because reducing size has been so central to everything we do. I don't think we need a bullet point to remind us.

P.S. I think @markelog meant [this pull](https://github.com/jquery/jquery/pull/2860).

**Comment by markelog** [comment]
> P.S. I think @markelog meant this pull.

Oh, right, sorry, fixex in my comment too.

**Comment by markelog** [comment]
> We will evaluate each of them with the criteria that we've always used; that is, based on size, performance, and readability.

btw, i think it would makes sense to document that in https://github.com/jquery/jquery/blob/master/CONTRIBUTING.md?

**Comment by timmywil** [comment]
> i think it would makes sense to document that

I'm not sure that's needed. It's pretty standard to judge pull requests based on those things. Besides, sometimes there's an X factor and we all vote.

**Comment by markelog** [comment]
mm, okay, but every project is different, like nowadays, ember/angular people for example, use transpilers, which is not that friendly to the size, whereas we consider size a priority. And comments like such 

> I didn't know that size was more important than performance

turned up couple times before.

**Comment by timmywil** [comment]
> I didn't know that size was more important than performance

This is certainly not a statement I would want to document, because it depends, right? In the case where the performance cost is too great, the size then becomes less important than performance. When is the performance cost too great? Is it some percentage? No. There maybe be other factors (like how much the performance cost will impact real use cases), but it is also subjective and subject to our opinions.

**Comment by markelog** [comment]
Isn't that our general position though? I think me and @gibson042 

> that we generally favor size over performance

were under that impression.

**Comment by markelog** [comment]
Anyway, i guess that is fruitless discussion, i don't feel strongly about this.
